### ISSUE TITLE
How to migrate from a simple expo repo to Nx Package-Based Monorepos

### ISSUE THREAD

**Issue by LeoAnt02** [xiongemi/nx-expo-monorepo#1]
Hello @xiongemi 

First and foremost, I'd like to extend my appreciation for your insightful articles and valuable contributions to the Nx Expo Ecosystem. Your work has been instrumental in guiding many of us in the community.

Regarding the existing documentation and resources, they primarily focus on initiating new projects using `nx g @nx/expo:app <my-app>` and are all Integrated Repositories. Could you shed some light on the specific reasons for this approach?

Our team has been managing an Expo project for over a year, and we are now exploring transitioning to a mono-repo setup, with a Package-Based Repo where each app maintains its own package.json for dependencies.(and the root is just defining the workspace)

Could you provide guidance on integrating an existing Expo project into an Nx Monorepo?

Created with
`npx create-expo-app --template`
(let's say we choose the template : File-based routing with TypeScript enabled) 

We've attempted to integrate our existing Expo project but no success yet,
Here's a brief overview of our approach so far:

TLDR : 

- We replicated the structure of a newly generated Nx Expo app in our existing Expo app, including the use of `eas-build-pre-install` and `eas-build-post-install` scripts in our package.json:p
-> Used the script and `tools/`  for eas  in our `package.json`
```
  "scripts": {
    "eas-build-pre-install": "cd ../../ && node tools/scripts/eas-build-pre-install.mjs . apps/my-app && cp package-lock.json apps/my-app",
    "eas-build-post-install": "cd ../../ && node tools/scripts/eas-build-post-install.mjs . apps/my-app"
  }
 ```
 -> created a similar `project.json` with all the Expo commands, including dependencies like `"ensure-symlink" and "sync-deps"`. 

 ```
       "dependsOn": [
        "ensure-symlink",
        "sync-deps"
      ],
```

However, we noticed this setup leaned towards an Integrated Repo structure, which we then removed. 
Are these dependencies crucial, or can we proceed without them?

#### Bugs on specific expo plugins 

Additionally, we've encountered several issues, mainely with custom plugins (expo)

- Bugs with Specific Expo Plugins: 
Running `nx run app:ios` results in a pod install failure due to plugin compatibility issues. 
The error points to CocoaPods not finding compatible versions for certain pods, requiring a higher minimum deployment target.
```
Cause: CocoaPods could not find compatible versions for pod`
Specs satisfying the `<package> (from `../../node_modules/<package>`)` dependency were found, but they required a higher minimum deployment target.
```

- EAS Build Failures: Similar to the above issue, remote cloud builds via EAS fail because of pod install failures or when applying the Script `'eas-build-pre-install'` :
Have this error: 
```
Failed to resolve plugin for module "expo-build-properties" relative to "/home/expo/workingdir/build/src"
```
(src being the workspace containing the expo project in this case)

- Nx Prebuild Issues: Prebuilds fail due to the absence of certain modules in my-app/node_modules, despite their presence in the root node_modules.

```
Error: Cannot find module './utils/autoAddConfigPlugins'
..../node_modules/@expo/cli/build/src/install/installAsync.js
..../node_modules/@expo/cli/build/src/install/index.js
..../node_modules/@expo/cli/build/bin/cli
```

I think we're close but we're missing something : 

## Questions : 
- Is there a particular rationale for the prevalent use of Integrated Repos in documentation and examples?
- Could you provide guidance on integrating an existing Expo project into an Nx Monorepo?
- What are the purposes of "ensure-symlink" and "sync-deps", and is it feasible to exclude them?
- Do you have any suggestions for resolving the plugin compatibility issue?
- Are there any naming conventions or structural requirements we should be aware of, such as the necessity to place our Expo app within an apps/ directory?

**Comment by xiongemi** [comment]
one thing i want to clarify, nx supports packaged based repo, but @nx/expo does not support packaged based repo by default.
what i meant is if you run `nx init` at workspace root, it should generate nx.json in the workspace, then you can run the commands like test and lint with nx cache using `nx run`.
For example, if in package.json scripts, you got a build script, instead of using `npm run build`, you can run `nx build <your package name>`.
however i don't think you can run `nx generate @nx/expo` in a package based repo.

---

Since your expo app is package-based repo, i think you can remove:
```
  "scripts": {
    "eas-build-pre-install": "cd ../../ && node tools/scripts/eas-build-pre-install.mjs . apps/my-app && cp package-lock.json apps/my-app",
    "eas-build-post-install": "cd ../../ && node tools/scripts/eas-build-post-install.mjs . apps/my-app"
  }
```

and

```
      "dependsOn": [
       "ensure-symlink",
       "sync-deps"
     ],
```
you don't need symlink and sync-deps for package-based repo.
I have a pr to remove ensure-symlink: https://github.com/nrwl/nx/pull/20896
sync-deps is mostly for integrated repo. since you are using packaged based, you do not need it.

---

i suspect why eas build failed because in the eas build steps, it only does an install inside your expo app only. it does not do an install at the workspace root.

maybe just add a script in your app's package.json, it should to your workspace root, and do an install there:
```
    "eas-build-pre-install": "cd .. && npm install"
```

---

i am not sure why pod install failed and i do not know what causes the plugin compatibility issue. do you have a repo?
since you run the prebuild command, i assume you are not using managed workflow, but a bare workflow?
is it not working on your local computer or eas cloud?
do you mean you got multiple pod versions in your workspace?

---

- Is there a particular rationale for the prevalent use of Integrated Repos in documentation and examples?
as i mentioned, all nx plugins do not support packaged based repo by default.


- Could you provide guidance on integrating an existing Expo project into an Nx Monorepo?
i assume you got a standable expo project and some other libraries outside the expo project. it is my approach:
  - run `npx create-nx-workspace --preset=npm` and it should create a packaged based repo.
  - it should create a `packages/` folder for you
  - then in the repo, under the packages folder, create a folder for the expo project, copy the expo project over.
  - try to run `nx graph` and make sure it is working
  - create the packages for the libraries and copy those
  - update package.json for expo and libraries
  - run `nx graph` to check
  - also run testing and lint command to make sure it is using nx cache `nx run-mang --target test lint`

- What are the purposes of "ensure-symlink" and "sync-deps", and is it feasible to exclude them?
i don't think you need it for package based repo

- Do you have any suggestions for resolving the plugin compatibility issue?
no idea what is the issue

- Are there any naming conventions or structural requirements we should be aware of, such as the necessity to place our Expo app within an apps/ directory?
i think you can placed anywhere in the workspace you want. if it is not found, just add it in the watchFolder in metro.config.js:
```
module.exports = withNxMetro(mergeConfig(defaultConfig, customConfig), {
  // Change this to true to see debugging info.
  // Useful if you have issues resolving modules
  debug: false,
  // all the file extensions used for imports other than 'ts', 'tsx', 'js', 'jsx', 'json'
  extensions: [],
  // Specify folders to watch, in addition to Nx defaults (workspace libraries and node_modules)
  watchFolders: ["packages", "path to your app"], // <- add it here
});
```

**Comment by LeoAnt02** [comment]
We've made significant progress and are nearly finished with our current phase. 
We're using Expo managed (we have no attachments to iOS and Android files)

Our approach so far has involved experimenting with symlinks and an integrated approach, primarily for mirroring purposes. (Will revisit a package-based repository structure in the future.)

Key solutions that facilitated our progress include:

- **Resolving Issues with npm Workspace**: 
Initially, including our Expo project in the root `package.json` workspace as part of an npm workspace presented numerous challenges. Removing the Expo project from the root `package.json` workspace resolved these issues. Interestingly, in creating a new workspace with NX, npm workspace is not utilized all.

  This change addressed the following error and allowed successful prebuilds:
  
  
```
  Error: Cannot find module './utils/autoAddConfigPlugins'
..../node_modules/@expo/cli/build/src/install/installAsync.js
..../node_modules/@expo/cli/build/src/install/index.js
..../node_modules/@expo/cli/build/bin/cli

```


- **Resolving Pod Install Issues**: 
The `pod install` issue, unrelated to Nx, was resolved by updating the target iOS build to a higher version number. This adjustment enabled successful EAS builds.

- **Metro Config Challenges**: 
We're currently encountering issues with the metro config. Although apps build successfully, errors occur upon opening. Utilizing tailwind CSS (nativewind) required us to add CSS to the blockList in `exclusionList`, which appears to resolve issues related to CSS being unable to resolve.

We're still facing issues related to resolving `node_modules`, as indicated by this error:

```
Unable to resolve "memoize-one" from "../node_modules/react-native/Libraries/Lists/FlatList.js"
``` 

Full error : 

```
Metro has encountered an error: While trying to resolve module `memoize-one` from file `/Users/leo/Documents/1_Code/project/node_modules/react-native/Libraries/Lists/FlatList.js`, the package `/Users/leo/Documents/1_Code/project/node_modules/memoize-one/package.json` was successfully found. However, this package itself specifies a `main` module field that could not be resolved (`/Users/leo/Documents/1_Code/project/node_modules/memoize-one/dist/memoize-one.cjs.js`. Indeed, none of these files exist:

  * /Users/leo/Documents/1_Code/project/node_modules/memoize-one/dist/memoize-one.cjs.js(.native|.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx|.ios.js|.native.js|.js|.ios.jsx|.native.jsx|.jsx|.ios.json|.native.json|.json|.ios.cjs|.native.cjs|.cjs)
  * /Users/leo/Documents/1_Code/project/node_modules/memoize-one/dist/memoize-one.cjs.js/index(.native|.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx|.ios.js|.native.js|.js|.ios.jsx|.native.jsx|.jsx|.ios.json|.native.json|.json|.ios.cjs|.native.cjs|.cjs): /Users/leo/Documents/1_Code/project/node_modules/metro/src/node-haste/DependencyGraph.js (289:17)

```


Here's our metro config
Any thought ?

```
const path = require('path');
const tailwind = require('tailwindcss/lib/cli/build');
const { withNxMetro } = require('@nx/expo');
const { getDefaultConfig } = require('@expo/metro-config');
const { mergeConfig } = require('metro-config');
const exclusionList = require('metro-config/src/defaults/exclusionList');

const defaultConfig = getDefaultConfig(__dirname);
const { assetExts, sourceExts } = defaultConfig.resolver;

/**
 * Metro configuration
 * https://facebook.github.io/metro/docs/configuration
 *
 * @type {import('metro-config').MetroConfig}
 */
const customConfig = {
  // transformer: {
  //   babelTransformerPath: require.resolve('react-native-svg-transformer'),
  // },
  resolver: {
    assetExts: assetExts.filter((ext) => ext !== 'svg'),
    sourceExts: [...sourceExts], //'svg'
    blockList: exclusionList([/^(?!.*\.(css|node_modules)).*\/dist\/.*/]),
    // unstable_enableSymlinks: true,
    // unstable_enablePackageExports: true,
  },
};

module.exports = (async () => {
  const config = withNxMetro(mergeConfig(defaultConfig, customConfig), {
    // Change this to true to see debugging info.
    // Useful if you have issues resolving modules
    debug: false,
    // all the file extensions used for imports other than 'ts', 'tsx', 'js', 'jsx', 'json'
    extensions: [],
    // Specify folders to watch, in addition to Nx defaults (workspace libraries and node_modules)
    watchFolders: [],
    // Enable CSS support.
    isCSSEnabled: true,
  });

  // Run Tailwind CLI to generate CSS files.
  await tailwind.build({
    '--input': path.relative(__dirname, './style.css'),
    '--output': path.resolve(__dirname, '../node_modules/.cache/expo/tailwind/eval.css'),
    '--watch': process.env.NODE_ENV === 'development' ? 'always' : false,
    '--poll': true,
  });

  return config;
})();
```

---

- Metro config seems quite different from the expo monorepo docs so we're a bit lost there 
Is there anything we're missing here ? babel config, tsconfig ? 

- On a side note : 
Is there a way to disable the Nx Console output wrapper in CI ? 
We rely on getting output, but Nx is adding a bunch of text that need cleaning in CI output
For exemple removing the 

```
 >  NX   --skip-nx-cache disables the connection to Nx Cloud for the current run.

   The remote cache will not be read from or written to during this run.


> nx run app:build -p android --profile staging --non-interactive --json

[
  {
    "id": "xxxxxxxx",
    ...
    
    
>  NX   Successfully ran target build for project app

```

**Comment by LeoAnt02** [comment]
After several attempts, we've successfully resolved all issues:

The solution was found within the metro configuration. The primary challenge involved integrating Tailwind (via Nativewind) with Expo 49 and Router, all within the context of an Nx workspace.

Immense thanks for all the assistance provided 🙏

Below is the final, effective metro configuration:

```
const path = require('path');
const tailwind = require('tailwindcss/lib/cli/build');
const { withNxMetro } = require('@nx/expo');
const { getDefaultConfig } = require('@expo/metro-config');
const { mergeConfig } = require('metro-config');
const exclusionList = require('metro-config/src/defaults/exclusionList');  

module.exports = (async () => {

const defaultConfig = getDefaultConfig(__dirname, {
  // Enable CSS support.
  isCSSEnabled: true,
});
const { assetExts, sourceExts } = defaultConfig.resolver;

/**
 * Metro configuration
 * https://facebook.github.io/metro/docs/configuration
 *
 * @type {import('metro-config').MetroConfig}
 */
const customConfig = {
  // transformer: {
  //   babelTransformerPath: require.resolve('react-native-svg-transformer'),
  // },
  resolver: {
    assetExts: assetExts.filter((ext) => ext !== 'svg'),
    sourceExts: [...sourceExts, 'jsx', 'js', 'ts', 'tsx', 'cjs', 'cjx', 'json', 'mjs', 'svg'], //'svg'
    blockList: exclusionList([/^(?!.*node_modules).*\/dist\/.*/]),
  // unstable_enableSymlinks: true,
    // unstable_enablePackageExports: true,
  },
};

  // Run Tailwind CLI to generate CSS files.
  await tailwind.build({
    '--input': path.relative(__dirname, 'style.css'),
    '--output': path.resolve(__dirname, 'node_modules/.cache/expo/tailwind/eval.css'),
    '--watch': process.env.NODE_ENV === 'development' ? 'always' : false,
    '--poll': true,
  });

  const config = withNxMetro(mergeConfig(defaultConfig, customConfig), {
    // Change this to true to see debugging info.
    // Useful if you have issues resolving modules
    debug: false,
    // all the file extensions used for imports other than 'ts', 'tsx', 'js', 'jsx', 'json'
    extensions: [],
    // Specify folders to watch, in addition to Nx defaults (workspace libraries and node_modules)
    watchFolders: [],
  });



  return config;
})();


```
